File: ios/ULinkReactNativeModule.swift
// ULinkReactNativeModule.swift
// Expo Module that wraps the ULink iOS SDK (ULinkSDK ~> 1.1.1).
//
// Design rules (from global-constraints.md):
//   - Module name: "ULinkReactNative"
//   - initialize() is async and idempotent; all calls queue until it resolves.
//   - Never wire dispose() to JS unmount / fast-refresh.
//   - Resolved cold-start/warm links reach JS via onDynamicLink / onUnifiedLink events,
//     not getInitialDeepLink().  Link delivery (AppDelegate forwarding) is Task 6.
//   - enableDeepLinkIntegration is always false on iOS (parseConfig enforces this).
//   - Client identity stays native sdk-ios — no SDK override needed.

import ExpoModulesCore
import ULinkSDK
import Combine

public class ULinkReactNativeModule: Module {

    // MARK: - State

    private var ulink: ULink?
    private var cancellables = Set<AnyCancellable>()
    private let queue = ULinkPendingQueue()
    private var initTask: Task<Void, Error>? = nil

    // MARK: - Module definition

    public func definition() -> ModuleDefinition {
        Name("ULinkReactNative")

        // ── Events ──────────────────────────────────────────────────────────
        Events("onDynamicLink", "onUnifiedLink", "onReinstallDetected", "onLog")

        // ── initialize ──────────────────────────────────────────────────────
        AsyncFunction("initialize") { (configMap: [String: Any], promise: Promise) in
            // If already initialised, resolve immediately (idempotent).
            if self.ulink != nil {
                promise.resolve()
                return
            }
            // If a concurrent init is in flight, queue behind it.
            if self.initTask != nil {
                let call = PendingCall.initialize(
                    config: configMap,
                    resolve: { promise.resolve() },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
                return
            }

            self.initTask = Task {
                do {
                    let config = try parseConfig(configMap)
                    let sdk = try await ULink.initialize(config: config)
                    self.ulink = sdk
                    self.subscribeStreams(sdk)
                    // Drain the method-call queue first so SDK event subscriptions
                    // are live before any buffered link is processed.
                    await self.queue.markReady(sdk, module: self)
                    // Drain any URLs buffered by the AppDelegate subscriber before
                    // this initialize() call completed (cold-start universal links,
                    // custom-scheme URLs).  processULinkUrl() emits on the Combine
                    // streams already wired above, so events reach JS correctly.
                    await ULinkIncomingLinkBuffer.shared.drain(sdk: sdk)
                    self.initTask = nil   // fix #6: clear task handle after successful init
                    promise.resolve()
                } catch {
                    self.initTask = nil
                    promise.reject("INITIALIZATION_ERROR", error.localizedDescription)
                }
            }
        }

        // ── createLink ──────────────────────────────────────────────────────
        AsyncFunction("createLink") { (paramsMap: [String: Any], promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    do {
                        let p = try parseParameters(paramsMap)
                        let resp = try await sdk.createLink(parameters: p)
                        promise.resolve(responseToMap(resp))
                    } catch {
                        promise.reject("CREATE_LINK_ERROR", error.localizedDescription)
                    }
                }
            } else {
                let call = PendingCall.createLink(
                    params:  paramsMap,
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── resolveLink ──────────────────────────────────────────────────────
        AsyncFunction("resolveLink") { (url: String, promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    do {
                        let resp = try await sdk.resolveLink(url: url)
                        promise.resolve(responseToMap(resp))
                    } catch {
                        promise.reject("RESOLVE_LINK_ERROR", error.localizedDescription)
                    }
                }
            } else {
                let call = PendingCall.resolveLink(
                    url:     url,
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── processULink ─────────────────────────────────────────────────────
        AsyncFunction("processULink") { (url: String, promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    guard let linkUrl = URL(string: url) else {
                        promise.reject("INVALID_URL", "Invalid URL: \(url)")
                        return
                    }
                    do {
                        let data = try await sdk.processULinkUrlThrowing(linkUrl)
                        if let data = data {
                            promise.resolve(resolvedDataToMap(data))
                        } else {
                            promise.resolve(nil as [String: Any?]?)
                        }
                    } catch {
                        promise.reject("PROCESS_ULINK_ERROR", error.localizedDescription)
                    }
                }
            } else {
                let call = PendingCall.processULink(
                    url:     url,
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── checkDeferredLink ────────────────────────────────────────────────
        AsyncFunction("checkDeferredLink") { (promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    do {
                        try await sdk.checkDeferredLinkAsync()
                        promise.resolve()
                    } catch {
                        promise.reject("DEFERRED_LINK_ERROR", error.localizedDescription)
                    }
                }
            } else {
                let call = PendingCall.checkDeferredLink(
                    resolve: { promise.resolve() },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getInitialDeepLink ───────────────────────────────────────────────
        AsyncFunction("getInitialDeepLink") { (promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    let data = await sdk.getInitialDeepLink()
                    if let data = data {
                        promise.resolve(resolvedDataToMap(data))
                    } else {
                        promise.resolve(nil as [String: Any?]?)
                    }
                }
            } else {
                let call = PendingCall.getInitialDeepLink(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getInitialUri ─────────────────────────────────────────────────────
        AsyncFunction("getInitialUri") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sdk.getInitialUrl()?.absoluteString)
            } else {
                let call = PendingCall.getInitialUri(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── setInitialUri ─────────────────────────────────────────────────────
        AsyncFunction("setInitialUri") { (uri: String, promise: Promise) in
            if let sdk = self.ulink {
                if let url = URL(string: uri) {
                    sdk.setInitialUrl(url)
                    promise.resolve()
                } else {
                    promise.reject("INVALID_URL", "Invalid URI: \(uri)")
                }
            } else {
                let call = PendingCall.setInitialUri(
                    uri:     uri,
                    resolve: { promise.resolve() },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getLastLinkData ───────────────────────────────────────────────────
        AsyncFunction("getLastLinkData") { (promise: Promise) in
            if let sdk = self.ulink {
                let data = sdk.getLastLinkData()
                if let data = data {
                    promise.resolve(resolvedDataToMap(data))
                } else {
                    promise.resolve(nil as [String: Any?]?)
                }
            } else {
                let call = PendingCall.getLastLinkData(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getInstallationId ─────────────────────────────────────────────────
        AsyncFunction("getInstallationId") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sdk.getInstallationId())
            } else {
                let call = PendingCall.getInstallationId(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getInstallationInfo ───────────────────────────────────────────────
        AsyncFunction("getInstallationInfo") { (promise: Promise) in
            if let sdk = self.ulink {
                if let info = sdk.getInstallationInfo() {
                    promise.resolve(installationInfoToMap(info))
                } else {
                    promise.resolve(nil as [String: Any?]?)
                }
            } else {
                let call = PendingCall.getInstallationInfo(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── isReinstall ───────────────────────────────────────────────────────
        AsyncFunction("isReinstall") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sdk.isReinstall())
            } else {
                let call = PendingCall.isReinstall(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getCurrentSessionId ───────────────────────────────────────────────
        AsyncFunction("getCurrentSessionId") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sdk.getCurrentSessionId())
            } else {
                let call = PendingCall.getCurrentSessionId(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── hasActiveSession ──────────────────────────────────────────────────
        AsyncFunction("hasActiveSession") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sdk.hasActiveSession())
            } else {
                let call = PendingCall.hasActiveSession(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── getSessionState ───────────────────────────────────────────────────
        AsyncFunction("getSessionState") { (promise: Promise) in
            if let sdk = self.ulink {
                promise.resolve(sessionStateString(sdk.getSessionState()))
            } else {
                let call = PendingCall.getSessionState(
                    resolve: { promise.resolve($0) },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── endSession ────────────────────────────────────────────────────────
        AsyncFunction("endSession") { (promise: Promise) in
            if let sdk = self.ulink {
                Task {
                    _ = await sdk.endSession()
                    promise.resolve()
                }
            } else {
                let call = PendingCall.endSession(
                    resolve: { promise.resolve() },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }

        // ── dispose ───────────────────────────────────────────────────────────
        // Advanced teardown only — do NOT call on component unmount.
        AsyncFunction("dispose") { (promise: Promise) in
            if let sdk = self.ulink {
                sdk.dispose()
                self.didDispose()
                promise.resolve()
            } else {
                let call = PendingCall.dispose(
                    resolve: { promise.resolve() },
                    reject:  { code, msg, _ in promise.reject(code, msg) }
                )
                Task { await self.queue.enqueue(call, module: self) }
            }
        }
    }

    // MARK: - Dispose state reset

    /// Resets all module-owned state after sdk.dispose() completes.
    /// Called from BOTH the direct dispose path and the queued dispose execution
    /// so a dispose-before-init can't leave stale state that collides with a
    /// later successful initialize().
    /// Internal (not private) so ULinkBridge.swift's queued .dispose execution can call it.
    func didDispose() {
        self.ulink = nil
        self.cancellables.removeAll()
        self.initTask = nil
    }

    // MARK: - Combine stream subscriptions

    private func subscribeStreams(_ sdk: ULink) {
        // Dynamic-link stream → onDynamicLink event
        sdk.dynamicLinkStream
            .receive(on: DispatchQueue.main)
            .sink { [weak self] data in
                self?.sendEvent("onDynamicLink", resolvedDataToMap(data))
            }
            .store(in: &cancellables)

        // Unified-link stream → onUnifiedLink event
        sdk.unifiedLinkStream
            .receive(on: DispatchQueue.main)
            .sink { [weak self] data in
                self?.sendEvent("onUnifiedLink", resolvedDataToMap(data))
            }
            .store(in: &cancellables)

        // Reinstall detection → onReinstallDetected event
        sdk.onReinstallDetected
            .receive(on: DispatchQueue.main)
            .sink { [weak self] info in
                self?.sendEvent("onReinstallDetected", installationInfoToMap(info))
            }
            .store(in: &cancellables)

        // Log stream → onLog event
        sdk.logStream
            .receive(on: DispatchQueue.main)
            .sink { [weak self] entry in
                self?.sendEvent("onLog", [
                    "level":     entry.level,
                    "tag":       entry.tag,
                    "message":   entry.message,
                    "timestamp": entry.timestamp,
                ])
            }
            .store(in: &cancellables)
    }
}

File: android/src/main/java/ly/ulink/reactnative/ULinkReactNativeModule.kt
package ly.ulink.reactnative

// ULinkReactNativeModule.kt
// Expo Module wrapping ly.ulink:ulink-sdk:1.1.0 for React Native / Expo.
//
// Design rules (global-constraints.md + task-7-brief.md):
//   - Name: "ULinkReactNative"
//   - initialize() is a suspend AsyncFunction; pre-init calls to other methods reject
//     with a clear error (requireSdk() throws). JS must await initialize() before other
//     calls — this is an intentional, documented Android/iOS behavioural difference.
//     iOS has a pendingCalls queue; Android does not (see Fix Report in task-7-report.md).
//   - enableDeepLinkIntegration is always forced false (ULinkBridge.config() enforces it).
//   - Cold-start links are buffered until BOTH: init complete AND first JS listener attached.
//   - Links are delivered via handleDeepLink(uri) → emits on SharedFlows → onDynamicLink /
//     onUnifiedLink events.  processULinkUri() is used only for the pull-method processULink().
//   - ALL mutations of pendingUris, initReady and observing are dispatched through `scope`
//     (Dispatchers.Main). This serialises them on a single thread — no locks needed, no race.
//   - dispose() cancels the module coroutine scope and clears SDK ref + buffer.
//     It does NOT call sdk.dispose() (which would cancel the SDK's own scope).
//   - Never wire dispose() to JS unmount / fast-refresh.

import android.content.Intent
import android.net.Uri
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ly.ulink.sdk.ULink

class ULinkReactNativeModule : Module() {

    // ── Coroutine scope ──────────────────────────────────────────────────────
    // All stream collectors and ALL mutations to the cold-start buffer run here.
    // Using Dispatchers.Main gives us a single-threaded serial executor for free —
    // no mutex / synchronized block needed for pendingUris, initReady, or observing.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // ── SDK reference (null until initialize() completes) ────────────────────
    // Written from scope (Main), read from Coroutine function bodies which may run
    // on the modulesQueue HandlerThread — @Volatile ensures cross-thread visibility.
    @Volatile private var sdk: ULink? = null

    // ── Two-gate cold-start buffer ───────────────────────────────────────────
    // A buffered link is only emitted once BOTH conditions hold:
    //   1. initReady  — ULink.initialize() has completed and streams are live.
    //   2. observing  — at least one JS event listener has attached (OnStartObserving).
    //
    // THREAD SAFETY: pendingUris, initReady and observing are accessed EXCLUSIVELY
    // through scope (Dispatchers.Main). Do NOT read or write them from any other
    // coroutine context or callback without dispatching through scope.launch { }.
    private var initReady = false
    private var observing = false
    private val pendingUris = mutableListOf<Uri>()

    // ── Module definition ────────────────────────────────────────────────────

    override fun definition() = ModuleDefinition {

        Name("ULinkReactNative")

        // ── Events ──────────────────────────────────────────────────────────
        Events("onDynamicLink", "onUnifiedLink", "onReinstallDetected", "onLog")

        // ── JS-listener gate ─────────────────────────────────────────────────
        // Fires on the Main thread when the first JS listener attaches.
        // Dispatched through scope to serialise with the buffer mutations in initialize().
        OnStartObserving {
            scope.launch {
                observing = true
                maybeFlush()
            }
        }

        // ── initialize ───────────────────────────────────────────────────────
        // Expo runs AsyncFunction Coroutine blocks on its modulesQueue HandlerThread,
        // NOT on the Main thread. ULink.initialize() is a suspend fun that blocks
        // until the SDK is ready — that is fine to run off Main.
        //
        // However, ALL mutations to pendingUris / initReady / observing must stay on
        // the Main thread. After the suspend call returns we hop back to scope (Main)
        // via withContext to do the launch-intent capture, set initReady, and flush.
        AsyncFunction("initialize") Coroutine { configMap: Map<String, Any?> ->
            // Idempotent: if already initialized return immediately.
            // sdk is @Volatile so this read is safe from any thread.
            if (sdk != null) return@Coroutine

            val ctx = appContext.reactContext?.applicationContext
                ?: throw IllegalStateException("Application context unavailable")

            val sdkConfig = ULinkBridge.config(configMap)

            // ULink.initialize() calls ProcessLifecycleOwner.get().lifecycle.addObserver()
            // which MUST run on the Main thread (Android lifecycle-2.x constraint).
            // Expo's AsyncFunction Coroutine runs on the modulesQueue (background thread),
            // so we must hop to Main for the entire ULink.initialize() call.
            val instance = withContext(Dispatchers.Main) { ULink.initialize(ctx, sdkConfig) }

            // Subscribe SharedFlows BEFORE setting initReady so collectors are live
            // when the first URI is flushed. subscribeStreams launches on scope (Main).
            subscribeStreams(instance)

            // Assign sdk before the Main-thread work so requireSdk() succeeds from
            // that point forward (visible cross-thread via @Volatile).
            sdk = instance

            // Hop to scope (Main) for ALL buffer/flag mutations and the flush.
            // withContext suspends the modulesQueue coroutine until this block finishes,
            // so initialize() doesn't return to the JS caller until the flush is done.
            withContext(scope.coroutineContext) {
                // Capture launch intent (cold-start deep link set before module existed).
                appContext.currentActivity?.intent?.data?.let { uri ->
                    bufferOrDeliver(uri)
                }
                // Open the init gate — maybeFlush() will fire if observing is already true.
                initReady = true
                maybeFlush()
            }
        }

        // ── createLink ───────────────────────────────────────────────────────
        AsyncFunction("createLink") Coroutine { paramsMap: Map<String, Any?> ->
            ULinkBridge.response(requireSdk().createLink(ULinkBridge.parameters(paramsMap)))
        }

        // ── resolveLink ──────────────────────────────────────────────────────
        AsyncFunction("resolveLink") Coroutine { url: String ->
            ULinkBridge.response(requireSdk().resolveLink(url))
        }

        // ── processULink — pull-method only, does NOT deliver via event ───────
        // Uses processULinkUri() which resolves-and-returns without emitting on
        // the SharedFlows. Link DELIVERY uses handleDeepLink().
        AsyncFunction("processULink") Coroutine { url: String ->
            ULinkBridge.resolvedData(requireSdk().processULinkUri(Uri.parse(url)))
        }

        // ── checkDeferredLink (non-suspend in SDK) ───────────────────────────
        AsyncFunction("checkDeferredLink") Coroutine { ->
            requireSdk().checkDeferredLink()
        }

        // ── getInitialDeepLink — suspend pull-method ─────────────────────────
        AsyncFunction("getInitialDeepLink") Coroutine { ->
            ULinkBridge.resolvedData(requireSdk().getInitialDeepLink())
        }

        // ── getInitialUri / setInitialUri ────────────────────────────────────
        AsyncFunction("getInitialUri") Coroutine { ->
            requireSdk().getInitialUri()?.toString()
        }

        AsyncFunction("setInitialUri") Coroutine { uri: String ->
            requireSdk().setInitialUri(Uri.parse(uri))
        }

        // ── getLastLinkData (non-suspend in SDK) ─────────────────────────────
        AsyncFunction("getLastLinkData") Coroutine { ->
            ULinkBridge.resolvedData(requireSdk().getLastLinkData())
        }

        // ── installation ─────────────────────────────────────────────────────
        AsyncFunction("getInstallationId") Coroutine { ->
            requireSdk().getInstallationId()
        }

        AsyncFunction("getInstallationInfo") Coroutine { ->
            ULinkBridge.installationInfo(requireSdk().getInstallationInfo())
        }

        AsyncFunction("isReinstall") Coroutine { ->
            requireSdk().isReinstall()
        }

        // ── session ───────────────────────────────────────────────────────────
        AsyncFunction("getCurrentSessionId") Coroutine { ->
            requireSdk().getCurrentSessionId()
        }

        AsyncFunction("hasActiveSession") Coroutine { ->
            requireSdk().hasActiveSession()
        }

        AsyncFunction("getSessionState") Coroutine { ->
            ULinkBridge.sessionState(requireSdk().getSessionState())
        }

        // endSession returns Boolean in the SDK but we return Unit to JS (matches iOS)
        AsyncFunction("endSession") Coroutine { ->
            requireSdk().endSession()
            Unit
        }

        // ── dispose ───────────────────────────────────────────────────────────
        // Cancels the module's coroutine scope and clears state.
        // Does NOT call sdk.dispose() (which would cancel the SDK's own lifecycle
        // scope — the SDK outlives any single module instance).
        AsyncFunction("dispose") Coroutine { ->
            scope.coroutineContext.cancelChildren()
            // Hop to Main to clear buffer state in the same serialised context.
            withContext(Dispatchers.Main) {
                sdk = null
                initReady = false
                observing = false
                pendingUris.clear()
            }
        }

        // ── OnNewIntent — warm deep links ─────────────────────────────────────
        // Fired on the Main thread when the app is already running and a new intent
        // arrives (warm link). Dispatch through scope to serialise with the buffer.
        OnNewIntent { intent: Intent ->
            scope.launch {
                intent.data?.let { uri -> bufferOrDeliver(uri) }
            }
        }
    }

    // ── Stream subscriptions ──────────────────────────────────────────────────

    private fun subscribeStreams(instance: ULink) {
        // dynamicLinkStream → onDynamicLink event
        instance.dynamicLinkStream
            .onEach { data ->
                sendEvent("onDynamicLink", ULinkBridge.resolvedData(data) ?: emptyMap<String, Any?>())
            }
            .launchIn(scope)

        // unifiedLinkStream → onUnifiedLink event
        instance.unifiedLinkStream
            .onEach { data ->
                sendEvent("onUnifiedLink", ULinkBridge.resolvedData(data) ?: emptyMap<String, Any?>())
            }
            .launchIn(scope)

        // onReinstallDetected → onReinstallDetected event
        instance.onReinstallDetected
            .onEach { info ->
                sendEvent(
                    "onReinstallDetected",
                    ULinkBridge.installationInfo(info) ?: emptyMap<String, Any?>()
                )
            }
            .launchIn(scope)

        // logStream → onLog event
        // timestamp is already a Long (epoch millis) in ULinkLogEntry — send as-is.
        instance.logStream
            .onEach { entry ->
                sendEvent(
                    "onLog",
                    mapOf(
                        "level" to entry.level,
                        "message" to entry.message,
                        "timestamp" to entry.timestamp,  // Long → JS number
                        "tag" to entry.tag
                    )
                )
            }
            .launchIn(scope)
    }

    // ── Cold-start / warm link buffer ────────────────────────────────────────
    // All functions below MUST be called from within scope (Dispatchers.Main).

    // Buffer the URI if either gate is still closed; otherwise deliver immediately.
    private fun bufferOrDeliver(uri: Uri) {
        if (initReady && observing) {
            deliverUri(uri)
        } else {
            pendingUris.add(uri)
        }
    }

    // Flush all buffered URIs when both gates open.
    // Snapshot-and-clear before any suspending call to avoid double-delivery.
    private fun maybeFlush() {
        if (!initReady || !observing) return
        val toDeliver = pendingUris.toList()
        pendingUris.clear()
        toDeliver.forEach { deliverUri(it) }
    }

    // Deliver a URI through handleDeepLink() which emits on the SharedFlows.
    // handleDeepLink() is non-suspend and launches its own coroutine internally.
    private fun deliverUri(uri: Uri) {
        sdk?.handleDeepLink(uri)
    }

    // ── SDK accessor ──────────────────────────────────────────────────────────

    // requireSdk() throws if called before initialize() completes. This is intentional:
    // Android requires JS to await initialize() before calling other methods.
    // Pre-init calls receive a clear rejection rather than silently queuing.
    // (iOS implements a pendingCalls queue; Android does not — see task-7-report.md Fix Report.)
    private fun requireSdk(): ULink =
        sdk ?: throw IllegalStateException(
            "ULink SDK not initialized. Call initialize() first and await it."
        )
}
